key-share/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+//! Key share of Threshold Signature Scheme (TSS)
+//!
+//! TSS protocols often share the same structure of key share. Having a separate crate with definition of the
+//! key share struct help reusing the code, keeping different implementations compatible and interopable.
+//!
+//! The crate provides [`DirtyCoreKeyShare`] that contains data such as: secret share, other signers commitments,
+//! public key and etc.
+//!
+//! [`DirtyCoreKeyShare`] may contain any data, not necessarily consistent. TSS protocol implementations typically
+//! don't want to handle inconsistent key shares and would rather assume that it's valid. [`Valid<T>`](Valid)
+//! is a type-guard stating that the value `T` it holds was validated. So, `Valid<DirtyCoreKeyShare>` (or
+//! [`CoreKeyShare`] type alias) can be used to express that only valid key shares are accepted.
+
+#![allow(non_snake_case)]
+#![deny(missing_docs, clippy::unwrap_used, clippy::expect_used, clippy::panic)]
+#![forbid(unused_crate_dependencies)]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+
+use generic_ec::{serde::CurveName, Curve, NonZero, Point, Scalar, SecretScalar};
+use generic_ec_zkp::polynomial::lagrange_coefficient;
+
+mod utils;
+mod valid;
+
+pub use self::valid::{Valid, Validate, ValidateError, ValidateFromParts};
+
+/// Core key share
+///
+/// Core key share is type alias to [`DirtyCoreKeyShare`] wrapped into [`Valid<T>`](Valid), meaning
+/// that the key share has been validated that:
+/// * Number of signers `n` doesn't overflow [`u16::MAX`], and that n >= 2
+/// * Signer index `i` is less than `n`
+/// * Signer public commitment matches the secret share
+/// * Threshold value is within range `2 <= t <= n`
+/// * All signers commitments sum up to public key
+///
+/// It's impossible to obtain [`CoreKeyShare`] for the key share that doesn't meet above requirements.
+///
+/// Only immutable access to the key share is provided. If you need to change content of the key share,
+/// you need to obtain dirty key share via [`Valid::into_inner`], modify the key share, and validate it
+/// again to obtain `CoreKeyShare`.
```

Comment:
I don't particularly care either way, but why did you decide to pull it from doc file to inline doc?

Reply:
Originally, I was using doc files to avoid repeating the same stuff in two different places. In this PR, I changed the docs so there's no repetition between `DirtyCoreKeyShare` and `CoreKeyShare`